GDJS/GDJS/Extensions/Builtin/CommonInstructionsExtension.cpp

```diff
@@ -101,6 +103,72 @@ CommonInstructionsExtension::CommonInstructionsExtension() {
         return outputCode;
       });
 
+  GetAllEvents()["BuiltinCommonInstructions::Async"].SetCodeGenerator(
+      [](gd::BaseEvent& event_,
+         gd::EventsCodeGenerator& codeGenerator,
+         gd::EventsCodeGenerationContext& parentContext) {
+        gd::AsyncEvent& event = dynamic_cast<gd::AsyncEvent&>(event_);
+
+        const gd::String callbackFunctionName =
+            "asyncCallback" +
+            gd::String::From(codeGenerator.GenerateSingleUsageUniqueIdFor(
+                &event.GetInstruction()));
+
+        // Generate callback code
+        gd::EventsCodeGenerationContext callbackContext;
+        callbackContext.AsyncInheritsFrom(parentContext);
+
+        // Generate actions
+        gd::String actionsCode = codeGenerator.GenerateActionsListCode(
+            event.GetActions(), callbackContext);
+
+        // Generate subevents
+        if (event.HasSubEvents())  // Sub events
+        {
+          actionsCode += "\n{ //Subevents\n";
+          actionsCode += codeGenerator.GenerateEventsListCode(
+              event.GetSubEvents(), callbackContext);
+          actionsCode += "} //End of subevents\n";
+        }
+
+        // Compose the callback function and add outside main
+        const gd::String actionsDeclarationsCode =
+            codeGenerator.GenerateObjectsDeclarationCode(callbackContext);
+        const gd::String callbackCode = "function " + callbackFunctionName +
+                                        "(runtimeScene, asyncObjectsList) {\n" +
+                                        actionsDeclarationsCode + actionsCode +
+                                        "}\n";
+        codeGenerator.AddCustomCodeOutsideMain(callbackCode);
+
+        // Generate code to backup the objects lists
+        gd::String objectsListsCode =
+            parentContext.IsAsync()
+                ? "asyncObjectsList = "
+                  "gdjs.LongLivedObjectsList.from(asyncObjectsList);\n"
+                : "const asyncObjectsList = new gdjs.LongLivedObjectsList();\n";
+        // The objects to be declared of the callback context tell all objects
+        // used in the callbacks so that they can be backed up now.
+        for (const gd::String& objectToBackup :
```

Comment:
Do you think there could be a way to save objects without having to know what lists of objects will be declared by all the callbacks "down the tree"?
Or do you think it's important to only save the lists that are both already declared (in any case, we do this) AND that we know will be declared in the future?

Reply:
It's important because of optimization. The long lived objects list needs to copy all objects references from the temporary objects list that could be used by other events, and create a function for each object for the callback. This operation is not the most expensive, but if we do it on objects we do not need to, this cost will increase fast. 

This ensures that only objects lists that will be used in the future in some way are actually added to the long lived list like this.

Reply:
I think it make sense. I would be keen to see if we can do this without storing asyncRoots for every context though :)